### PR TITLE
Core/Spell: Remove all raid auras on party leave

### PR DESCRIPTION
**Pull request by robinsch**
**Changes proposed**:
- Remove all raid auras on party leave (https://youtu.be/S2znqLUQZzo?t=1m28s)
- As of Patch 2.1, "... In addition, the soulstone buff will now be removed if the target or caster leaves the party or raid." 

**Target branch(es)**: 335/6x

**Tests performed**: (Does it build, tested in-game, etc)
Tested in-game

**Known issues and TODO list**:
- [x] Find a better way

### CODE REVIEW

**Comment by Aokromes** [comment]
IMHO this needs generic fix for all buffs.

**Comment by robinsch** [comment]
Which other buffs?

**Comment by Shauren** [comment]
things like arcane intellect, for example

**Comment by robinsch** [comment]
Arcane Intellect? Why should this be removed when leaving a party? It can be casted on non-party members.

**Comment by Keader** [comment]
Focus Magic dont need be a group too

File: src/server/game/Groups/Group.cpp [hunk 1]
```diff
@@ -484,14 +485,38 @@ bool Group::RemoveMember(ObjectGuid guid, const RemoveMethod& method /*= GROUP_R
 
     sScriptMgr->OnGroupRemoveMember(this, guid, method, kicker, reason);
 
+    // As of Patch 2.1, "The soulstone buff will now be removed if the target or caster leaves the party or raid."
+    Player* player = ObjectAccessor::FindConnectedPlayer(guid);
+    if (player)
```
Comment (Keader) [inline src/server/game/Groups/Group.cpp]: why not ```if (Player* player = ObjectAccessor::FindConnectedPlayer(guid))```  ?

**Comment by Shauren** [comment]
I don't know about the case when caster leaves party, but if you leave the group while having a raid buff on you cast by old group member it gets removed

**Comment by robinsch** [comment]
Oh yes - https://youtu.be/S2znqLUQZzo?t=1m28s I will take a look at it.

**Comment by robinsch** [comment]
```c++
    if (player)
    {
        Unit::AuraMap const& ownedAuras = player->GetOwnedAuras();
        for (Unit::AuraMap::const_iterator itr = ownedAuras.begin(); itr != ownedAuras.end(); ++itr)
        {
            Aura* aura = itr->second;

            if (aura->GetSpellInfo()->IsRaidMemberTarget() && aura->GetCasterGUID() != guid)
                player->RemoveAura(aura);
        }

        for (member_witerator witr = m_memberSlots.begin(); witr != m_memberSlots.end(); ++witr)
        {
            if (Player* groupMember = ObjectAccessor::FindConnectedPlayer(witr->guid))
            {
                Unit::AuraMap const& memberAuras = groupMember->GetOwnedAuras();
                for (Unit::AuraMap::const_iterator itr = memberAuras.begin(); itr != memberAuras.end(); ++itr)
                {
                    Aura* aura = itr->second;
                    if (aura->GetSpellInfo()->IsRaidMemberTarget() && aura->GetCasterGUID() == guid)
                    {
                        groupMember->RemoveAura(aura);
                    }
                        
                }
            }
        }
```

There should be a better way of checking if aura should be removed or not. I have no idea how to put that into "better" C++ code. 

For removed player we should check all auras with IsRaidMemberTarget() and casterGUID != his guid.
For all other group members we must do the same check but for casterGUID == removed player guid.

Is there a way to do these kind of checks in a nicer way (because they almost look the same.)

**Comment by Temppus** [comment]
Maybe some raid spell attribute ?

**Comment by P-Kito** [comment]
https://github.com/TrinityCore/TrinityCore/pull/16848/commits/8de7e12a6e16dd9e38deafe683e2165a26125410#diff-7f4f8fd9e7fa96bc9c9b16415f0e6e7bR4172

RemoveOwnedAuras RemoveAppliedAuras looks cleaner I think^^

**Comment by robinsch** [comment]
I've check all related attributes and did not find them having something common, so I'd rather check it by target flag.

File: src/server/game/Spells/SpellInfo.cpp [hunk 2]
```diff
@@ -1131,6 +1131,14 @@ bool SpellInfo::IsAllowingDeadTarget() const
...
 }
 
+bool SpellInfo::IsRaidMemberTarget() const
+{
+    for (uint8 i = 0; i < MAX_SPELL_EFFECTS; ++i)
+        if (Effects[i].IsEffect() && (Effects[i].TargetA.GetTarget() & (TARGET_UNIT_CASTER_AREA_RAID | TARGET_UNIT_TARGET_RAID | TARGET_UNIT_TARGET_AREA_RAID_CLASS)))
+            return true;
+    return false;
+}
+
```
Comment (boxa) [inline src/server/game/Spells/SpellInfo.cpp]: more:
```
bool SpellMgr::IsGroupBuff(SpellEntry const* spellInfo)
{
    for (uint8 i = 0; i < MAX_EFFECT_INDEX; ++i)
    {
        switch (spellInfo->EffectImplicitTargetA[i])
        {
            case TARGET_SINGLE_PARTY:
            case TARGET_ALL_PARTY_AROUND_CASTER:
            case TARGET_ALL_PARTY:
            case TARGET_ALL_PARTY_AROUND_CASTER_2:
            case TARGET_AREAEFFECT_PARTY:
            case TARGET_ALL_RAID_AROUND_CASTER:
            case TARGET_AREAEFFECT_PARTY_AND_CLASS:
                if (IsPositiveEffect(spellInfo, SpellEffectIndex(i)) &&
                    (spellInfo->Effect[i] == SPELL_EFFECT_APPLY_AURA ||
                     spellInfo->Effect[i] == SPELL_EFFECT_APPLY_AREA_AURA_PARTY ||
                     spellInfo->Effect[i] == SPELL_EFFECT_APPLY_AREA_AURA_RAID))
                {
                    return true;
                }
                break;
            default:
                break;
        }
    }

    // some custom cases
    switch (spellInfo->Id)
    {
        case 72590:                             // Fortitude (triggered by AoE spell with ScriptEffect)
            return true;
        default:
            break;
    }

    return false;
}
```

Comment (Shauren) [inline src/server/game/Spells/SpellInfo.cpp]: @boxa forgive me for being so blunt but in my opinion spell targeting and aura stacking rules in mangos are complete garbage with the amount of hacks "exclusions" it needs for various things

On topic: you can try using SpellImplicitTargetInfo::GetCheckType - TARGET_CHECK_PARTY, TARGET_CHECK_RAID, TARGET_CHECK_RAID_CLASS,

**Comment by Aokromes** [comment]
I think it was all positive buffs, i don't remember exactly, i will dig youtube.
https://youtu.be/KxNpa2TY4Uk?t=1m8s

**Comment by robinsch** [comment]
Updated

File: src/server/game/Spells/SpellInfo.cpp [hunk 3]
```diff
@@ -1131,6 +1131,15 @@ bool SpellInfo::IsAllowingDeadTarget() const
     return HasAttribute(SPELL_ATTR2_CAN_TARGET_DEAD) || Targets & (TARGET_FLAG_CORPSE_ALLY | TARGET_FLAG_CORPSE_ENEMY | TARGET_FLAG_UNIT_DEAD);
 }
 
+bool SpellInfo::IsGroupBuff() const
+{
+    for (uint8 i = 0; i < MAX_SPELL_EFFECTS; ++i)
+        if (Effects[i].TargetA.GetCheckType() & (TARGET_CHECK_PARTY | TARGET_CHECK_RAID | TARGET_CHECK_RAID_CLASS))
```
Comment (Shauren) [inline src/server/game/Spells/SpellInfo.cpp]: check type isnt a flag, use switch

Comment (robinsch) [inline src/server/game/Spells/SpellInfo.cpp]: wooops my bad didnt check

**Comment by Kittnz** [comment]
@robinsch are there any other needed changes for this to be finished?

**Comment by robinsch** [comment]
No

**Pull request merged**